### PR TITLE
[SPARK-32693][SQL] Compare two dataframes with same schema except nullable property

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This PR changes key data types check in `HashJoin` to use `sameType`.

### Why are the changes needed?

Looks at the resolving condition of `SetOperation`, it requires only each left data types should be `sameType` as the right ones. Logically the `EqualTo` expression in equi-join, also requires only left data type `sameType` as right data type. Then `HashJoin` requires left keys data type exactly the same as right keys data type, looks not reasonable.

It makes inconsistent results when doing `except` between two dataframes.

If two dataframes don't have nested fields, even their field nullable property different, `HashJoin` passes the key type check because it checks field individually so field nullable property is ignored.

If two dataframes have nested fields like struct, `HashJoin` fails the key type check because now it compare two struct types and nullable property now affects. 

### Does this PR introduce _any_ user-facing change?

Yes. Making consistent `except` operation between dataframes.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
cc @cloud-fan @maropu

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala [hunk 1]
```diff
@@ -106,7 +106,10 @@ trait HashJoin extends BaseJoinExec with CodegenSupport {
   }
 
   protected lazy val (buildKeys, streamedKeys) = {
-    require(leftKeys.map(_.dataType) == rightKeys.map(_.dataType),
+    require(
+      leftKeys.map(_.dataType)
+        .zip(rightKeys.map(_.dataType))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala:111]: `zip` allows different length, we should check it explicitly.

**Review by cloud-fan — APPROVED** [review]
makes sense to me

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala (hunk 1 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala:111]: ok

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
Thanks, @viirya ! Merged to master. Could you backport the fix into branch-3.0/2.4? The fix has some conflicts with them.

**Comment by viirya** [comment]
Thanks @cloud-fan @maropu.

I will backport this later.

**Pull request closed without merging**